src/fparser/tests/test_parsefortran.py

```diff
@@ -0,0 +1,142 @@
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+##############################################################################
+'''
+Tests the fparser.parsefortran module.
+'''
+
+import logging
+import pytest
+import fparser.parsefortran
+import fparser.readfortran
+
+
+@pytest.fixture
+def log():
+    logger = logging.getLogger('fparser')
+    log = fparser.tests.logging_utils.CaptureLoggingHandler()
+    logger.addHandler(log)
+    yield log
+    logger.removeHandler(log)
+
+
+def test_log_empty(log):
+    '''
+    Tests that a reader without next() method causes an event to be logged.
+    '''
+    class EmptyReader(object):
+        '''
+        A faux reader with no next() method.
+        '''
+        id = 'thingumy'
+
+    unit_under_test = fparser.parsefortran.FortranParser(EmptyReader())
+    unit_under_test.analyze()
+    assert log.messages == {'debug':    [],
+                            'info':     ['Nothing to analyze.'],
+                            'warning':  [],
+                            'error':    [],
+                            'critical': []}
+
+
+def test_log_cache(log):
+    '''
+    Tests that using a chached reader object logs an event.
+    '''
+    class Readerlike(object):
+        '''
+        Dummy reader class, the only purpose of which is to have an id and not
+        cause the parser to fail.
+        '''
+        id = 'thisun'
+
+        def next(self):
+            '''
+            Simple non-failure-causing method.
+            '''
+            yield 'NOT A THING'
+            yield 'MODULE foo_mod'
+            raise StopIteration
+
+    # Expect everything to go okay, no log messages.
+    log.reset()
+    unit_under_test = fparser.parsefortran.FortranParser(Readerlike())
+    assert log.messages == {'debug':    [],
+                            'info':     [],
+                            'warning':  [],
+                            'error':    [],
+                            'critical': []}
+
+    # This time we should use a cached log.
+    unit_under_test = fparser.parsefortran.FortranParser(Readerlike())
+    assert log.messages == {'debug':    [],
+                            'info':     ['using cached thisun'],
+                            'warning':  [],
+                            'error':    [],
+                            'critical': []}
+
+
+def test_log_failure(log):
+    '''
+    Tests that an unexpected read failure causes an event to be logged.
+    '''
+    class FailyReader(fparser.readfortran.FortranStringReader):
+        '''
+        Modifies the FortranStringReader class to raise an exception from
+        the next() method.
+        '''
+        def __init__(self):
+            '''
+            Default constructor.
+            '''
+            super(FailyReader, self).__init__('The end')
+
+        def next(self, ignore_comments=False):
+            '''
+            Failing method.
+            '''
+            raise Exception('That''s all folks!')
```

Comment:
It seems not. All the examples of monkeypatching I had seen used a lambda, which can't raise exceptions. I suppose it might raise a "divide by zero". Anyway, I re-read the documentation and tried using a function instead and that works.